### PR TITLE
refactor mo_sethet

### PR DESCRIPTION
**Pull request by tangshuaiqi**
separate out a few small subroutines. re-organize some code and variables.

Also removed all unused chemical species in current MAM4, only hno3 (used in H2SO4 calculation), h2o2 and so2 left.

### CODE REVIEW

**Review by jeff-cohere — APPROVED** [review]

File: components/eam/src/chemistry/mozart/mo_sethet.F90 [hunk 1]
```diff
@@ -296,109 +272,41 @@ subroutine sethet( het_rates, press, zmid,  phis, tfld, &
...
-       else
-          xsogx(:ncol,k)  = 0._r8
-       end if
-       if( spc_so2_ndx > 0 ) then
-          xso2(:ncol,k)  = qin(:ncol,k,spc_so2_ndx) * xhnm(:ncol,k)
-       else
-          xso2(:ncol,k)  = 0._r8
-       end if
-    end do
+
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/mozart/mo_sethet.F90:291]: The `ktop_all` calculation looks problematic to me as it is taking the `minval` of the columns. The number of columns and their grouping depends on the PE layout. I am surprised that this code doesn't produce NBFB answers for different PE layouts (PET test). We can do nothing here as we like to keep the code as is.

File: components/eam/src/chemistry/mozart/mo_sethet.F90 [hunk 2]
```diff
@@ -573,311 +372,224 @@ subroutine sethet( het_rates, press, zmid,  phis, tfld, &
...
-             het_rates(i,k,hcooh_ndx) = max( rain(i,k) / (h2o_mol*(work1(i) + 1._r8/(xhen_ch3cooh(i,k)*work2(i)))),0._r8 )
-          endif
-          if ( hcn_ndx > 0 ) then
-             het_rates(i,k,hcn_ndx     ) = max( rain(i,k) / (h2o_mol*(work1(i) + 1._r8/(xhen_hcn(i,k)*work2(i)))),0._r8 )
-          endif
-          if ( ch3cn_ndx > 0 ) then
-             het_rates(i,k,ch3cn_ndx   ) = max( rain(i,k) / (h2o_mol*(work1(i) + 1._r8/(xhen_ch3cn(i,k)*work2(i)))),0._r8 )
+             work3(icol) = satf_hno3 *  max( rain(icol,kk) / &
+                           (h2o_mol*(work1(icol) + 1._r8/(xhen_hno3(icol,kk)*work2(icol)))), 0._r8 )
+             het_rates(icol,kk,h2so4_ndx) =  work3(icol) + tmp_hetrates(icol,kk,1)
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/mozart/mo_sethet.F90]: Calculation of work3 and het_rates look exactly the same for different species, can we factor it out?

**Review by singhbalwinder — APPROVED** [review]
Looks great! I have suggested to factor out a repeated calc if it is doable.

File: components/eam/src/chemistry/mozart/mo_sethet.F90 (hunk 2 above)
Comment (tangshuaiqi) [reply components/eam/src/chemistry/mozart/mo_sethet.F90]: yes, refactored it out.

**Review by singhbalwinder — APPROVED** [review]

**Comment by tangshuaiqi** [comment]
All tests passed. merge this PR

comp_testall_sethet_mam4_org_v2_baselines: 9 tests
  PASS ERP_P64x1.ne4pg2_oQU480.F2010.compy_intel
  PASS ERS.ne4pg2_oQU480.F2010.compy_intel
  PASS PEM_P64x1.ne4pg2_oQU480.F2010.compy_intel
  PASS PET.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D_Ln5_P1x1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D_Ln5_P32x1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_Ly1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS.ne30pg2_EC30to60E2r2.F2010.compy_intel

**Pull request merged**